cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinCondition.java

```diff
@@ -0,0 +1,224 @@
+      if (schema1 == null) {
+        return;
+      }
+      while (keyIter.hasNext()) {
+        JoinKey otherKey = keyIter.next();
+        Schema otherSchema = stageMap.get(otherKey.getStageName()).getSchema();
+        if (otherSchema == null) {
+          continue;
+        }
+        Iterator<String> fieldIter = otherKey.getFields().iterator();
+        for (String key1Field : key1.getFields()) {
+          // guaranteed to be non-null because of earlier validation
+          Schema key1Schema = schema1.getField(key1Field).getSchema();
+          key1Schema = key1Schema.isNullable() ? key1Schema.getNonNullable() : key1Schema;
+          Schema.Type key1Type = key1Schema.getType();
+
+          // don't need to check hasNext() because already verified they are all the same length
+          String otherField = fieldIter.next();
+          Schema otherFieldSchema = otherSchema.getField(otherField).getSchema();
+          otherFieldSchema = otherFieldSchema.isNullable() ? otherFieldSchema.getNonNullable() : otherFieldSchema;
+          Schema.Type otherType = otherFieldSchema.getType();
+
+          if (key1Type != otherType) {
+            throw new InvalidJoinConditionException(String.format(
+              "Type mismatch on join keys. '%s'.'%s' is of type '%s' while '%s'.'%s' is of type '%s'.",
+              key1.getStageName(), key1Field, key1Schema.getDisplayName(),
+              otherKey.getStageName(), otherField, otherFieldSchema.getDisplayName()));
+          }
+        }
+      }
+    }
+
+    /**
+     * Builds an OnKeys condition.
+     */
+    public static class Builder {
+      private final Set<JoinKey> keys;
+      private boolean dropNullKeys;
+
+      private Builder() {
+        this.keys = new HashSet<>();
+        this.dropNullKeys = false;
+      }
+
+      public Builder addKey(JoinKey key) {
+        this.keys.add(key);
+        return this;
+      }
+
+      public Builder setKeys(Collection<JoinKey> keys) {
+        this.keys.clear();
+        this.keys.addAll(keys);
+        return this;
+      }
+
+      public Builder setShouldDropNullKeys(boolean dropNullKeys) {
+        this.dropNullKeys = dropNullKeys;
+        return this;
+      }
+
+      public OnKeys build() {
+        if (keys.size() < 2) {
+          throw new InvalidJoinConditionException("Must specify a join key for each input stage.");
+        }
+        Map<Integer, Set<String>> numFieldsToStages = new HashMap<>();
+        for (JoinKey joinKey : keys) {
+          int numFields = joinKey.getFields().size();
+          Set<String> stages = numFieldsToStages.computeIfAbsent(numFields, k -> new HashSet<>());
+          stages.add(joinKey.getStageName());
+        }
+        // this means there are stages with different number of fields.
+        // it's the equivalent of trying to join on A.id = B.id and A.name = [null]
+        if (numFieldsToStages.size() > 1) {
+          Iterator<Integer> keyIter = numFieldsToStages.keySet().iterator();
```

Comment:
I think it's better to iterate through the `numFioeldsToStages` to use all stages to form the error message. E.g. if the map is like this:

```
{
  1: ["s1"],
  2: ["s2"],
  3: ["s3"]
}
```
It's better to show all the stages with the number of join fields that they declared.